### PR TITLE
Working "beavis-ci" manual CIT system

### LINKED ISSUE #188: Simple manual CIT and output deployment for tutorial notebooks
This is something I started doing over in the stack club, and I think we need it here too: a script that runs all our notebooks, makes static html pages out of them, and deploys them to an orphan branch so that they can be viewed, with plots (while the notebooks remain lightweight and easily diffed). I'll copy over the [relevant script](https://github.com/lsst-com/notebooks/blob/master/beavis-ci.csh) and see if I can make it work.

### PR DESCRIPTION
**Pull request by drphilmarshall**
I think the `beavis-ci.csh` script in the Notebooks folder is now in good shape:
* When the `beavis-ci.csh` script is run in the Notebooks folder, all the tutorial notebooks are rendered as notebooks with outputs and pushed to an orphan "rendered" branch, along with the logs from `jupyter nbconvert`
* The [Notebooks folder README file](https://github.com/LSSTDESC/DC2_Repo/blob/issue/188/clean/Notebooks/README.md) contains an index table with links to the notebook, its rendering (viewed on nbviewer.com to avoid confusion), and its log file. 
* Most notebooks do not run successfully: some debugging may be required.

I think this closes #188 - but see what you think!

PS. Ignore the FoF matching notebook diff - I thought I could use it as a test, but it was too broken...

### CODE REVIEW

**Comment by wmwv** [comment]
This seems reasonable.  Can you remove the changes to the FoF matching notebook?

**Comment by drphilmarshall** [comment]
Done!

File: Notebooks/README.md [hunk 1]
```diff
@@ -13,7 +13,13 @@ validation programs.
...
-| DC2 Coadd Run1.1p Butler Access HSC selection | Use the Butler to apply the HSC object selection cuts | [ipynb](Notebooks/DC2_Coadd_Run1.1p_Butler_Access_HSC_selection.ipynb), [html](http://htmlpreview.github.io/?https://github.com/LSSTDESC/DC2_Repo/blob/html/Notebooks/DC2_Coadd_Run1.1p_Butler_Access_HSC_selection.html) | [Jim Chiang](https://github.com/LSSTDESC/DC2_Repo/issues/new?body=@jchiang87) |
+| DC2 Reference Catalog Reader | Simple example of reading the reference catalog | [ipynb](DC2%20Reference%20Catalog%20Reader.ipynb), [rendered](https://nbviewer.jupyter.org/github/LSSTDESC/DC2_Repo/blob/rendered/Notebooks/DC2_Reference_Catalog_Reader.nbconvert.ipynb), [log](https://github.com/LSSTDESC/DC2_Repo/blob/rendered/Notebooks/log/DC2_Reference_Catalog_Reader.log)  | [Anze Slosar](https://github.com/LSSTDESC/DC2_Repo/issues/new?body=@slosar) |
```
[the remark this answers is not in the document]
Comment (cwwalter) [reply Notebooks/README.md]: Recently I was working on a more complicated table than this for the imSim wiki in markdown.  Since every row of the table has to be a single line just connected with "|" it is really hard to edit parse when you are adding long text.

What I found worked better in this case was to write just this page in reStructured text instead.  GH parses those files too.

Here is the page for reference.  
https://github.com/LSSTDESC/imSim/wiki/Feature-Matrix

A table entry looks like this:

```
.. list-table::
   :widths: 10 10 10 15 15
   :header-rows: 1

   * - Effect
     - Implementation
     - Data / Model Source
     - Short description
     - Validation Page and Notebooks

   * - Brighter Fatter
     - GalSim Feature (link)
     - BF slopes measured on TS-X See LCA-XXX
     - GalSim reads in output from full electrostatic solver and does linear interpolation,  See page link
     - `Page link <https://github.com/LSSTDESC/imSim/wiki/brighter_fatter_validation>`_
       The solution to a full electrostatic solver interpolates distorted pixel boundaries as
       function of pixel level. These boundaries are used as the sensor is being filled with electrons.
```

Which is a lot easier for a human to edit.  You might want to consider doing this.  It would just take a few minutes to change I think.

File: Notebooks/beavis-ci.csh [hunk 2]
```diff
@@ -93,26 +93,29 @@ git clone git@github.com:LSSTDESC/DC2_Repo.git
 cd DC2_Repo/Notebooks
 
 if ($html) then
-    echo "Making static HTML pages from the available notebooks:"
+    echo "Making static HTML pages from the available master branch notebooks:"
```
Comment (cwwalter) [inline Notebooks/beavis-ci.csh]: Adding some documentation on whether someone needs to run this manually and push it or if it is a GH hook/app somehow would be useful.

I finally had to bite the bullet and change my shell from tcsh to bash and I feel your pain, but we probably should get this converted to bash at some point.  The reason (I was shocked to learn!) is that tcsh isn't even installed by default on most linux distributions.  So, it might not be there for a lot of people.

**Comment by wmwv** [comment]
When I run this myself on Cori I get the same successful and broken Notebooks as @drphilmarshall .

One note to add the README.md is that Notebooks must completely compile successfully.  E.g., the DC2 static coadd reader Notebook has a cell that's known not to work; it's the last cell, and it's fine when running the Notebook, but breaks auto-generation.  I think it's fine to require the Notebooks to compile, but this means we can't have, e.g., pedagogically-motivated intentionally broken cells.

(For full disclosure, the broken cell in the static coadd reader is not pedagogically intentional; it's something I don't know how to do right.)

**Comment by drphilmarshall** [comment]
Thanks @wmwv - I added notes to the README as you suggested, [see what you think.](https://github.com/LSSTDESC/DC2_Repo/blob/issue/188/clean/Notebooks/README.md)  It's odd that you cannot push, I would expect you to be able to since you have Write access through the SSim team. Is your token under-powered, somehow?

**Comment by wmwv** [comment]
Could you document what rights the token needs to have?  There are lots of boxes one can check when creating a token.

**Comment by wmwv** [comment]
Please rebase to the latest master before merging.

**Comment by wmwv** [comment]
And minor:
You might update the README to add the additional motivation that we clear outputs for committing not just to avoid "repo bloat", but to avoid conflicts on every run due to semantically equivalent but bitwise-distinct output blobs.

**Comment by yymao** [comment]
Not sure if this is the right place to mention this, but at some point we need to sync the [DC2 Data Product Overview confluence page](https://confluence.slac.stanford.edu/x/oJgHDg) with the presentation in this repo.

**Comment by wmwv** [comment]
@drphilmarshall Could you kindly rebase to master before merging?
There are a lot of merges in this branch history that are somewhat distracting.

If you prefer, I'm happy to do the rebase.

**Comment by drphilmarshall** [comment]
Thanks all!

@EiffL It was a real token until GitHub detected it in the script header and cancelled it... I've replaced it with `$GITHUB_API_TOKEN` now

@cwwalter That rst table is much nicer, I agree. It's kind of fragile though - I switched to rst and started trying to make the table (just with two entries to start with), but am stuck. [Can you see the problem](https://github.com/LSSTDESC/DC2_Repo/edit/issue/188/clean/Notebooks/README.rst)?

@wmwv @cwwalter I switched to bash, and tested the resulting [beavis-ci.sh script](https://github.com/LSSTDESC/DC2_Repo/edit/issue/188/clean/Notebooks/beavis-ci.sh). I've _almost_ got a cron job going...

@wmwv I added your note on semantically-equivalent output blob conflict to the README

**Comment by cwwalter** [comment]
> @cwwalter That rst table is much nicer, I agree. It's kind of fragile though - I switched to rst and started trying to make the table (just with two entries to start with), but am stuck. Can you see the problem?

Sorry... in another meeting now. I can look later. I found the main thing that caused the table to break/not render is if the text does is not aligned properly.  I mean if (For example) text in the column that you wrapped in the editor doesn't also start at the right indent level.

**Comment by cwwalter** [comment]
@drphilmarshall First problem is the line in the first row of the table that starts Anze is indented one space too far.  Fix that and the table will render.

**Comment by drphilmarshall** [comment]
@wmwv The `beavis-ci.sh` script now seems to be fully functional, and I have set up a cron job to run it every 6 hours. I think this PR is ready to merge: I attempted a `git rebase master` but it seems to have created a significantly messier history than it had before (and I had to fix conflicts with multiple commits). See what you think - hopefully you can just hit the big green button. Please :-)

**Comment by yymao** [comment]
@drphilmarshall several tutorial notebooks are missing in the table:

- [ ] [HOD](https://github.com/LSSTDESC/DC2_Repo/blob/master/Notebooks/protoDC2%20HOD.ipynb)
- [ ] [mass relations](https://github.com/LSSTDESC/DC2_Repo/blob/master/Notebooks/protoDC2%20mass%20relations.ipynb)
- [ ] [N(z)](https://github.com/LSSTDESC/DC2_Repo/blob/master/Notebooks/N_vs_z_distributions.ipynb)
- [ ] [truth catalog](https://github.com/LSSTDESC/DC2_Repo/blob/master/Notebooks/DC2_truth_example.ipynb)
- [ ] [coadd part 3](https://github.com/LSSTDESC/DC2_Repo/blob/master/Notebooks/DC2%20Coadd%20Run1.1p%20GCR%20tutorial%20--%20Part%20III:%20Guided%20Challenges.ipynb)

And "reference catalog" should be removed from the table.

**Comment by wmwv** [comment]
I'll be happy to clean up the history.  Give me an hour.

**Comment by wmwv** [comment]
@drphilmarshall In the future, please don't merge master back into your branch.  This is why the history is really complicated and hard to rebase.  Is there a particular button in the UI you're using that encourages merging master into the branch?

**Comment by yymao** [comment]
@wmwv When a repo has continuous integration set up and a branch is out of date, GitHub shows a button to merge master back to the branch.

**Comment by cwwalter** [comment]
@wmwv Oooh.. quick comment..

> In the future, please don't merge master back into your branch. This is why the history is really complicated and hard to rebase. Is there a particular button in the UI you're using that encourages merging master into the branch?

I recently went through trying to rebase master onto a long lived branch branch and then push it to back to my PR. It's pretty confusing and I usually mess up one of the steps.  I think I now understand what to do after going through it a few times, but this is definitely something that we should help people out with.  It's pretty non-obvious. Even what you are doing on a basic level is non-obvious and then people need to understand they should "push -f" and never do that to master, understand how to use rebase --continue etc.

So, we should tell people why it is so much better but then also have a something to point them at.  Just trying to figure out git stuff on the web usually makes you want to smash something :)

**Comment by drphilmarshall** [comment]
@wmwv I'm using command line git, no app or UI, and following the instructions as printed by git. This time I did not do `git merge master`, but instead I did `git rebase master` and then fixed every conflict it threw as it replayed all the history (that is, I edited the conflicted files, then did `git add`, and then did `git rebase --continue` as advised). When this was all done, I did `git push` without the `-f`. It sounds like this might have been my biggest mistake. After going through the whole procedure, TBH I am left unconvinced that `git rebase` is a good idea, for two reasons. 1) It's re-writing history (never a good idea?), and 2) it seems to take a lot more developer time than `git merge` does.  Is a "clean" history really worth the extra effort? Once I understand all this I'll be happy to add it to my understanding in the GettingStarted repo: I agree with @cwwalter that this is not an easy operation (I found it hard to follow the `git` documentation too)...

@yymao Thanks for pointing out the missing notebooks. I don't know why they did not appear in my dev branch when I rebased onto master - probably because of something I did wrong in the rebase, I expect. I can add them to the README anyway, though - by default, the `beavis-ci` script now reads the README to see what it should test.

BTW my cron job fails with a bunch of "jupyter: command not found" errors, so I've turned it off and submitted a ticket. The NERSC help doesn't have much to say on running cron on cori, which modules to load etc and I have run out of time for trial and error!

**Comment by yymao** [comment]
This 'rebase vs. merge' thing is like 'vim vs. emacs' which you can find endless debate on the Internet :) While I do like clean history too, personally I think it's better for our collaboration to use merge because most people are not git experts and merge is usually safer. Just my 2 cents.

**Comment by cwwalter** [comment]
FWIW the project uses rebase.  It really does make it easier to understand the history.

**Comment by cwwalter** [comment]
> When this was all done, I did git push without the -f. It sounds like this might have been my biggest mistake. After going through the whole procedure, TBH I am left unconvinced that git rebase is a good idea, for two reasons. 1) It's re-writing history (never a good idea?), and 2) it seems to take a lot more developer time than git merge does. Is a "clean" history really worth the extra effort? Once I understand all this I'll be happy to add it to my understanding in the GettingStarted repo: I agree with @cwwalter that this is not an easy operation (I found it hard to follow the git documentation too)...

Interesting.  Basically it you have rewritten the history I thought it basically won't let you do the push.  I found you are required to use the -f.  In terms of rewriting history, yes but the idea is that you are just doing this only to a PR.  It is part of working on and getting your code in shape and no one else should be affected by it.

**Comment by wmwv** [comment]
@drphilmarshall If you didn't do a `git push -f` then the GitHub version of the branch didn't get updated and we don't see it.

**Comment by wmwv** [comment]
Because I'm writing on the Project side I'm clearly used to doing things the Project way.
I understand the choice `astropy` made to not rebase and why they did it.

But merging master back to one's branch multiple times, and then merging that branch back to master is really confusing.  It creates multiple copies of each commit that has been done on master.  Even if we end up going with the merge without rebase, I would plead that we not merge master to branch.  Thanks, @yymao , for explaining why this easily happens in the GitHub setup.

**Comment by wmwv** [comment]
> BTW my cron job fails with a bunch of "jupyter: command not found" errors, so I've turned it off and submitted a ticket. The NERSC help doesn't have much to say on running cron on cori, which modules to load etc and I have run out of time for trial and error!

I would guess that your cron job runs as just at bare login without sourcing the DESC DM Stack install.

**Comment by cwwalter** [comment]
> I would plead that we not merge master to branch. Thanks, @yymao , for explaining why this easily happens in the GitHub setup.

If you have a branch and you need to get it up to date with master and you don't rebase or merge it to master what is the other option? I'm not following...

**Comment by drphilmarshall** [comment]
@cwwalter I think Michael's saying that we should just merge this PR without merging `master` into `issue/188/clean` first. ("clean", hah)

OK, so when I now try to rebase to master it's asking me to redo all the conflict fixes that I already did, so I gave up. I guess if I had done `git push -f` things would be different?

Meanwhile, the [history according to GitHub network graph](https://github.com/LSSTDESC/DC2_Repo/network) actually looks OK, so I'm going to vote we just merge this PR and I promise to do better on the rebasing next time. 

I think you're right about the cron job environment, @wmwv . I don't see any desc stuff in my .bashrc, .bashrc.ext, .bash_profile or .bash_profile.ext so would not know what to have cron source to set up my environment in full. Seems odd that `jupyter` is not available by default, though. I'll let you know what NERSC tell me.

**Comment by wmwv** [comment]
> If you have a branch and you need to get it up to date with master and you don't rebase or merge it to master what is the other option? I'm not following...

You merge your branch *to* master when you're done.

**Comment by wmwv** [comment]
Anyway, since I've been the one complaining, that also means it's on me to be willing to do the work to achieve the utopia I dream of.

I'm rebasing now and will merge when I'm done.

**Comment by wmwv** [comment]
Thanks for the indulgence, @drphilmarshall

**Comment by drphilmarshall** [comment]
Thanks @wmwv . I have to say, that commit history is a thing of beauty :-)
